common/src/sql/dbinit.sql

```diff
@@ -128,6 +139,29 @@ CREATE INDEX ON omicron.public.service (
     sled_id
 );
 
+/*
+ * A table describing resource usage which may be associated
+ * with a collection of objects, including:
+ * - Projects
+ * - Organizations
+ * - Silos
+ * - Fleet
+ */
+CREATE TABLE omicron.public.resource_usage (
+    /* Should match the UUID of the corresponding collection */
+    id UUID PRIMARY KEY,
+
+    -- The amount of physical disk space which has been provisioned
+    -- on behalf of the collection.
+    physical_disk_bytes_provisioned INT8 NOT NULL,
```

Comment:
The work in this PR seems to be to be duplicating the existing Crucible resource accounting that was put in with #1706, except that here it's being modified by adding diffs instead of being computed idempotently, and it's not directly tied to volume creation and deletion. For example, there are cases where deleting a snapshot at the Nexus level does not delete the allocated regions on disk.

I haven't been in any of the conversations about resource usage, but in my mind it would be safer to compute disk bytes being used on demand when asked (through `GET /system/metrics/{resource_name}`) instead of keeping this `resource_usage` table around. Said another way, I do see the value in being able to answer user queries about resource usage per [project, org, silo, fleet], but I don't see the value in a persistent table.

Reply:
If we don't use a table, re-computing these values will incur a full-table scan -- doing so will require inspecting the rows of every single disk / instance in the fleet each time the value is queried.

Reply:
I think this might depend on whether this is used vs. allocated and whether it's virtual or physical.

Say we're talking about virtual disk bytes allocated.  I think we only need that for reporting purposes, not correctness.  It already exists in the database as SUM(size) for all disks, images, snapshots, etc. (I know it's not quite that simple, but the data must be in the database, I think.)  There are at least three ways I can see solving this:

- Create a denormalized table, as we've done here.  Pros: cost is amortized across the operations that would modify the data, and the data is always up-to-date.  But a big con: it's easy for the data to become wrong (it just takes one errant code path, execute one time) and you'd potentially never even know it was wrong.
- Poll on-demand -- i.e., do the computation on-demand when asked.  I agree with the concern that this won't scale well.
- Run a periodic batch job that computes these values and stores them, maybe directly into Clickhouse.  Big pros: high confidence that the data is correct, and any inaccuracy can be fixed in the next run.  Cons: data is only available on a frequency limited by how long the job takes to run.  (I suspect this is not _at all_ a problem for the intended use case.)  These are often a source of continuing maintenance.
- Do the denormalized table, but also run a periodic batch job to detect inaccuracies.  This is probably the most expensive option in terms of engineering time, but it combines some of the best of these: the data is always up-to-date, we can identify and correct inaccuracies, and if the job starts taking longer to run, that only affects how quickly we fix inaccuracies, not how frequently fresh data becomes available.  If we do this, any inaccuracies should be flagged and we should try to root-cause them (good luck).  This is a bit hard to implement because the underlying data can change as we're validating it.  Maybe we could use CockroachDB's `AS OF SYSTEM TIME` to operate on a snapshot, calculate the delta, and apply the delta later to the running table.

On the other hand, the choice might be pretty different if we're talking about a metric that the control plane needs for correctness.  I think we've been assuming that the control plane needs a view of space allocated and free in order to correctly fail provisions when insufficient space is available.  This would push towards option (1) and I think it'd behoove us to consider how we will either ensure that no code path ever gets it wrong or (more likely) deal with the inevitable drift in accuracy.

But I also wonder if this is fundamentally not the right approach to resource allocation in the control plane.  For resources like disk space, it's impossible to reliably determine the physical resources that will be needed for a given virtual resource. There are potentially significant costs to making an estimate and having it be wrong (whichever direction you choose).  Might another approach be to choose a low watermark of available physical capacity and stop allocating user resources if we're below that mark?  This has some of the same problems as trying to estimate it on a per-allocation basis, but I think with this approach we can be more conservative to avoid the risk of _under_estimating space used, and the cost of over-estimating (which is wasted space) can be amortized across all allocations.  A key point here is that I think a production system is very unlikely to actually run out.  It needs to _handle_ running out, of course, but an operator is likely going to either buy more storage or attempt to reduce utilization _before_ reaching the "out of space" point.  So it doesn't matter as much if we waste a little space below the low watermark (i.e., if we overestimate it).  This can even overlap with the buffer we've said we want to avoid zpools getting too full.  But as I think about this more, I'm not sure this gets you out of the business of estimating future utilization because if you have a zillion allocations for disk space but nobody's using it, you'll never hit the watermark even though you should definitely stop allocating.  Thoughts on any of this?

Reply:
I tried to do a similar comparison here: https://github.com/oxidecomputer/omicron/issues/1734#issuecomment-1265737111

When I was considering doing this for a "fleet-wide" view, I was most inclined to use a batch job - it's just a single `SUM`, so even if it needs to do a full table scan, at least it would be *one* full-table scan. However, to actually sum by project, organization, silo, and fleet requires a bit more complexity - either a CTE that emits data proportional to every single collection in the fleet (which could be a fairly large output) or it would require multiple full table scans.

This is what made me significantly more inclined towards the de-normalized view. I'm happy to periodically validate these values (that also has the nice property that we can validate data which is churning, and largely ignore unchanged rows - plus, it should be entirely optional).

On the note of "whether or not this is the right approach to resource allocation in the control plane" - I want to be really explicit about what this PR does and does not do. My priority was surfacing collection-level aggregate views to an operator for **visibility**, not necessarily to impose constraints. That seems like a very natural follow-up to this work, but I 100% agree that taking a closer look at "physical, not virtual" resource usage seems like a necessary part of that story.

Frankly, I don't know how we'll set capacity for user-level resources. RFD 70 indicates that we should not be over-provisioning, but if that's the case, our 3x redundancy by each crucible region will significantly inhibit the amount of storage that can be used on a rack, even if physical utilization is low. We have options in this space, but again, my goal in this PR is *merely to surface information, rather than act on it*.